Add doc for vSphere CSI Migration (#21444)
kubernetes/website: content/en/docs/concepts/storage/volumes.md

More examples can be found [here](https://github.com/kubernetes/examples/tree/master/staging/volumes/vsphere).

#### CSI Migration

> {{< feature-state for_k8s_version="v1.19" state="beta" >}}

Reviewer: @xing-yang 
we are going with `alpha` as per https://github.com/kubernetes/kubernetes/pull/90911#discussion_r435529744
Once test results are posted, we will generate another PR to bump to `beta`

Author: Sure

Author: Done

The CSI Migration feature for vsphereVolume, when enabled, shims all plugin operations

> from the existing in-tree plugin to the `csi.vsphere.vmware.com` {{< glossary_tooltip text="Container Storage Interface" term_id="csi" >}} Driver. In order to use this feature, the [vSphere CSI

Reviewer 2: nit:
```suggestion
from the existing in-tree plugin to the `csi.vsphere.vmware.com` {{< glossary_tooltip text="CSI" term_id="csi" >}} Driver. In order to use this feature, the [vSphere CSI
```

Author: Done

Driver](https://github.com/kubernetes-sigs/vsphere-csi-driver)
must be installed on the cluster and the `CSIMigration` and `CSIMigrationvSphere`
features must be enabled.

#### CSI Migration Complete
{{< feature-state for_k8s_version="v1.19" state="beta" >}}

> To turn off the vsphereVolume plugin from being loaded by controller manager and kubelet, you need to set this feature flag to true. This requires `csi.vsphere.vmware.com` {{< glossary_tooltip text="Container Storage Interface" term_id="csi" >}} driver being installed on all worker nodes.

Reviewer 2: I'm not sure what this section is explaining.

Author: This actually follows the similar doc section from the Azure driver.
It means that when the "CSI Migration Complete" is set true, CSI migration will stop working.  This requires CSI driver to be installed on all nodes.

More examples can be found [here](https://github.com/kubernetes/examples/tree/master/staging/volumes/vsphere).

#### CSI Migration

{{< feature-state for_k8s_version="v1.19" state="beta" >}}

The CSI Migration feature for vsphereVolume, when enabled, shims all plugin operations
from the existing in-tree plugin to the `csi.vsphere.vmware.com` {{< glossary_tooltip text="CSI" term_id="csi" >}} Driver. In order to use this feature, the [vSphere CSI

> Driver](https://github.com/kubernetes-sigs/vsphere-csi-driver)

Reviewer 3: Min vsphere + hardware versions, and known features that won't work should be documented here as well

More examples can be found [here](https://github.com/kubernetes/examples/tree/master/staging/volumes/vsphere).

#### CSI Migration

{{< feature-state for_k8s_version="v1.19" state="beta" >}}

The CSI Migration feature for vsphereVolume, when enabled, shims all plugin operations
from the existing in-tree plugin to the `csi.vsphere.vmware.com` {{< glossary_tooltip text="CSI" term_id="csi" >}} Driver. In order to use this feature, the [vSphere CSI
Driver](https://github.com/kubernetes-sigs/vsphere-csi-driver)
must be installed on the cluster and the `CSIMigration` and `CSIMigrationvSphere`
features must be enabled. This also requires minimum vSphere vCenter/ESXi Version to be 7.0u1 and minimum HW Version to be VM version 15.

> The following StorageClass parameters from vsphereVolume plugin will not be supported by vSphere CSI driver:

Reviewer: Can we mention below for clarification?

Existing PVs created using these parameters will be migrated to vSphere CSI driver, but new Volumes cannot be created honoring these parameters

Author: Good point

More examples can be found [here](https://github.com/kubernetes/examples/tree/master/staging/volumes/vsphere).

> #### CSI Migration

Reviewer 2: nit (OK to merge without fixi)
```suggestion
#### CSI migration
```

Author: Done

> {{< feature-state for_k8s_version="v1.19" state="beta" >}}

Reviewer 2: Should this say alpha or beta?

Author: It is beta.

> {{< feature-state for_k8s_version="v1.19" state="beta" >}}

Reviewer 3: The feature gate page also needs to be updated

Author: Done

The CSI Migration feature for vsphereVolume, when enabled, shims all plugin operations

> from the existing in-tree plugin to the `csi.vsphere.vmware.com` {{< glossary_tooltip text="CSI" term_id="csi" >}} Driver. In order to use this feature, the [vSphere CSI

Reviewer 4: nit: lowercase, `driver`?

Reviewer 4: nit: The feature defaults to `enabled`? The cluster requirements are listed below.
Could you remove `when enabled`?

Reviewer 3: This is unlike other beta features and actually defaults to off.

Author: Done

Driver](https://github.com/kubernetes-sigs/vsphere-csi-driver)
must be installed on the cluster and the `CSIMigration` and `CSIMigrationvSphere`

> features must be enabled. This also requires minimum vSphere vCenter/ESXi Version to be 7.0u1 and minimum HW Version to be VM version 15.

Reviewer 2: ```suggestion
[feature gates](/docs/reference/command-line-tools-reference/feature-gates/) must be enabled.

This also requires minimum vSphere vCenter/ESXi Version to be 7.0u1 and minimum HW Version to be VM version 15.
```

Author: Done

> The following StorageClass parameters from vsphereVolume plugin will not be supported by vSphere CSI driver:

Reviewer 4: The following StorageClass parameters from the vsphereVolume plugin are not supported by the vSphere CSI driver:

Author: Fixed

* diskformat
* hostfailurestotolerate
* forceprovisioning
* cachereservation
* diskstripes
* objectspacereservation
* iopslimit

> Existing volumes created using these parameters will be migrated to the vSphere CSI driver, but new volumes created by the vSphere CSI driver will not be honoring these parameters.

Reviewer 4: nit: must be migrated to the vSphere CSI driver OR is a process handling the migration?
nit: will not honor these parameters

Author: I think "will be" is correct in the sentence.  We are talking about what will happen with CSI migration.

> The following StorageClass parameters from vsphereVolume plugin will not be supported by vSphere CSI driver:
> 
> * diskformat
> * hostfailurestotolerate
> * forceprovisioning
> * cachereservation
> * diskstripes
> * objectspacereservation
> * iopslimit
> 
> Existing volumes created using these parameters will be migrated to the vSphere CSI driver, but new volumes created by the vSphere CSI driver will not be honoring these parameters.

Reviewer 2: nit: I'd make this into a note shortcode

_beware_ that this might not preview correctly until commit 847f57ecf53b46bc7ab5e134b1e78ee2113fce21 is merged into [_dev-1.19_](https://github.com/kubernetes/website/tree/dev-1.19). You could cherry-pick that locally to verify that it's looking how you intend, then omit it from the branch you push.

```suggestion
{{< note >}}
The following StorageClass parameters from the built-in vsphereVolume plugin are not supported by the vSphere CSI driver:

* `diskformat`
* `hostfailurestotolerate`
* `forceprovisioning`
* `cachereservation`
* `diskstripes`
* `objectspacereservation`
* `iopslimit`

Existing volumes created using these parameters will be migrated to the vSphere CSI driver, but new volumes created by the vSphere CSI driver will not be honoring these parameters.
{{< /note >}}
```

Author: Looks like commit 847f57e is merged into master.  My PR will be merged into dev-1.19.  Probably should make this a separate PR when that commit is backported to dev-1.19?

Reviewer 2: _master_ will be merged into _dev-1.19_ to incorporate that kind of change (a regular thing). It's fine to do this as a separate PR after that happens.

(It's also fine IMO to merge in a note shortcode that looks OK now but will render even better with changes from 847f57ecf53b46bc7ab5e134b1e78ee2113fce21)

Author: Sure.  I added it as a note.

#### CSI Migration Complete
{{< feature-state for_k8s_version="v1.19" state="beta" >}}

To turn off the vsphereVolume plugin from being loaded by controller manager and kubelet, you need to set this feature flag to true. This requires `csi.vsphere.vmware.com` {{< glossary_tooltip text="CSI" term_id="csi" >}} driver being installed on all worker nodes.

> 

Reviewer 2: What does this mean? I wasn't sure, especially the heading.
Bear in mind that this is conceptual documentation - would it make sense to add a [blog](https://kubernetes.io/blog/) post about how to migrate, rather than include this section here?

Reviewer 3: This is a separate feature gate that will disable the loading of the legacy driver entirely. The difference is subtle but there is some fallback logic in the implementation that will handle the scenario where the nodes are still running the legacy driver. Turning on the migration complete feature gate disables that fallback logic.

Reviewer 3: Also this section is mostly copied form the wording that we use for the other csi migration features. If you think the wording is confusing, we should work on updating the wording for all the cloud providers.

Reviewer 2: OK, I've logged https://github.com/kubernetes/website/issues/22521 and am happy given that to skip fixing it here.
